### DISCUSSION TITLE
zh/algorithms/hanoitower

### DISCUSSION THREAD

**Discussion by giscus** [selfboot/ai_gallery#11] [Q&A]
# zh/algorithms/hanoitower

经典的汉诺塔问题，从任意初始状态开始，用递归算法找到后续步骤。可以调整圆盘数量，选择手动或自动模式，获取提示，切换模式，源码开放。

https://gallery.selfboot.cn/zh/algorithms/hanoitower

**Comment by Hongshi0622** [comment]
感恩，让我学会了递归

**Comment by coder-Song** [comment]
虽然1个圆盘和2个圆盘很简单，但觉得还是有必要设置出来的，可以帮助形成思维

**Reply by coder-Song** [reply]
如果能标上序号就更好了，要不然圆盘多了有时候看不出来长短不明显

**Reply by selfboot** [reply]
嗯嗯，这两个确实可以优化下，我抽时间改改，感谢

**Reply by selfboot** [reply]
这里现在支持最少 2 个盘子，每个盘子上也有数字了

**Comment by ffdffdfggg** [comment]
几分钟就弄出来了，成就感满满

**Comment by nonworry** [comment]
总结一下就是，需要移动的那一摞是单数时，就把最顶上圆盘往目标位置移动；是双数时，就把圆盘往目标之外的位置移动。

**Comment by Spencer-Chen-China** [comment]
自动模式如果有暂停功能更好理解

**Reply by Spencer-Chen-China** [reply]
手动模式可以加个上一步，防止走错了只能重开

**Comment by kerosene530** [comment]
很好用

**Reply by Wuduquan** [reply]
代码在那里查看呢

**Reply by selfboot** [reply]
https://github.com/selfboot/ai_gallery/blob/main/src/app/%5Blang%5D/algorithms/hanoitower/content.js 这里，每个页面标题右边那里有代码链接

**Comment by zezepro** [comment]
移动一个n阶 等效于 移动一次n-1阶 然后移动最下面那个 再移动一次n-1阶 即
A_n = 2A_{n-1} + 1

**Comment by SEDT38** [comment]
感恩，让我学会了递归
